SqlClient instrumentation to support Microsoft.Data.SqlClient EventSource

mbakalov opened the pull request:
Fixes #1597.

## Changes

* `SqlClientInstrumentationOptions` now has different public API on .NET Core and .NET Framework
* Made `SqlClientDiagnosticListener` .NET Core-only (it was referencing the options properties no longer available on .NET Framework)
  * For that had to refactor some common properties into a new `SqlActivitySourceHelper`
  * DiagnosticListener is no longer created/used under .NET Framework
* On .NET Framework the new `SetStatementText` is **false by default**
* SqlEventSourceListener is now listening to both event sources ("Microsoft-AdoNet-SystemData" and "Microsoft.Data.SqlClient.EventSource")
* Updated Microsoft.Data.SqlClient NuGet in the Test project from 1.1.1 to 2.1.0

## Background

Recently, `Microsoft.Data.SqlClient` renamed their EventSource from "Microsoft-AdoNet-SystemData" to "Microsoft.Data.SqlClient.EventSource". The `System.Data.SqlClient` still uses the old event source, so the instrumentation needs to listen to both event sources.

TODO:

* [x] unit-tests, including "fake event source" ones to have a 2nd fake source
* [x] behavior change: new EventSource reports CommandText not only for StoredProcedures now but for Text statements as well
* [x] comments and README updates  
* [x] `CHANGELOG.md` updated for non-trivial changes
~~* [ ] Changes in public API reviewed - shouldn't be any public API changes~~
* [ ] Changes in public API reviewed - there are public API changes for `SetStatementText`

cijothomas reviewed (commented):
LGTM. Please add an entry in chagelog.md.
We need to update tests to use mds as well rigth?

mbakalov commented:
> 
> 
> LGTM. Please add an entry in chagelog.md.
> We need to update tests to use mds as well rigth?

@cijothomas - thanks, yes, working on that still, will push later today. The semantics of the new EventSource are different a bit too (it now captures CommandText always, not just for SPs) so I want to make sure that is supported properly as well.

mbakalov commented:
Actually, annoyingly that behavior is different enough that it is going to result in confusing API.

`System.Data.SqlClient` basically does this:
```csharp
private void WriteBeginExecuteEvent()
{
    if (SqlEventSource.Log.IsEnabled() && Connection != null)
    {
        string commandText = CommandType == CommandType.StoredProcedure ? CommandText : string.Empty;
        SqlEventSource.Log.BeginExecute(GetHashCode(), Connection.DataSource, Connection.Database, commandText);
    }
}
```

While MDS does this:
```csharp
private void WriteBeginExecuteEvent()
{
    if (Connection != null)
    {
        SqlClientEventSource.Log.BeginExecute(GetHashCode(), Connection.DataSource, Connection.Database, CommandText);
    }
}
```

So MDS always reports `CommandText` which is great, but now in our listener we cannot tell if we are dealing with an SP or a Text statement. Previously we could say "CommandText is non-empty so it has to be an SP", now we can't.

This becomes confusing because the "options" has two separate settings for those: `SetTextCommandContent` and `SetStoredProcedureCommandName` (currently `SetTextCommandContent` does nothing on `System.Data.SqlClient` netfx).

To clean this up, I am suggesting the following in the `SqlClientInstrumentationOptions` API, basically have only one toggle that does both (but only on netfx).  Any thoughts?
```csharp

        // .NET Framework implementation uses SqlEventSource from which we can't reliably distinguish
        // StoredProcedures from regular Text sql commands.
#if NETFRAMEWORK

        /// <summary>
        /// Gets or sets a value indicating whether or not the <see cref="SqlClientInstrumentation"/> should
        /// add the text of the executed Sql commands as the <see cref="SemanticConventions.AttributeDbStatement"/> tag.
        /// Default value: False.
        /// </summary>
        public bool SetStatementText { get; set; }
#else
        /// <summary>
        /// Gets or sets a value indicating whether or not the <see cref="SqlClientInstrumentation"/> should add the names of <see cref="CommandType.StoredProcedure"/> commands as the <see cref="SemanticConventions.AttributeDbStatement"/> tag. Default value: True.
        /// </summary>
        public bool SetStoredProcedureCommandName { get; set; } = true;

        /// <summary>
        /// Gets or sets a value indicating whether or not the <see cref="SqlClientInstrumentation"/> should add the text of <see cref="CommandType.Text"/> commands as the <see cref="SemanticConventions.AttributeDbStatement"/> tag. Default value: False.
        /// </summary>
        public bool SetTextCommandContent { get; set; }
#endif
```

mbakalov commented:
Alternatively, the instrumentation can set `db.statement` on netfx MDS only if _both_ `SetStoredProcedureCommandName` and `SetTextCommandContent` are set to true. Less code and less ifdefs in the codebase (especially tests), but might be more confusing to users?

cijothomas commented:
> Alternatively, the instrumentation can set `db.statement` on netfx MDS only if _both_ `SetStoredProcedureCommandName` and `SetTextCommandContent` are set to true. Less code and less ifdefs in the codebase (especially tests), but might be more confusing to users?

Can we instead use separate listeners for MDS and SDS? Then listener know exactly if its getting sp or sqltext. 
Sorry haven't fully read the code. but an initial thought.

CodeBlanch commented on src/OpenTelemetry.Instrumentation.SqlClient/Implementation/SqlEventSourceListener.netfx.cs:
```diff
@@ -55,7 +63,13 @@ protected override void OnEventSourceCreated(EventSource eventSource)
+                this.EnableEvents(eventSource, EventLevel.Informational, (EventKeywords)1);
+            }
+
+            if (eventSource?.Name.StartsWith(MdsEventSourceName, StringComparison.Ordinal) == true)
```
You could probably make this an `else if` as it will never match both clauses.

mbakalov commented:
> 
> 
> > Alternatively, the instrumentation can set `db.statement` on netfx MDS only if _both_ `SetStoredProcedureCommandName` and `SetTextCommandContent` are set to true. Less code and less ifdefs in the codebase (especially tests), but might be more confusing to users?
> 
> Can we instead use separate listeners for MDS and SDS? Then listener know exactly if its getting sp or sqltext.
> Sorry haven't fully read the code. but an initial thought.

Yeah that is the issue I think. Even if MDS was a separate listener, it just gets a `CommandText` and at that point has no way of telling if it is sp or sqltext.

CodeBlanch commented:
@mbakalov Bummer about the change to command text by MS! I like your idea to change the options.

mbakalov commented:
Introduced a separate `SetStatementText` property just on .NET Framework (had to refactor code just a bit to get there) - please take a look.

It actually turned out that the reason net461 tests were failing was only somewhat related to the renaming of the EventSource. The test project referenced MDS nuget version 1.1.1, which actually still uses the old EventSource name. :) The tests were flaky for a reason similar to what was being discussed in https://github.com/elastic/apm-agent-dotnet/pull/704#discussion_r381478027: when running both `System.Data.SqlClient` and `Microsoft.Data.SqlClient` tests in the same session, the `OnEventSourceCreated` was firing multiple types during a single test, overwriting the `eventSource` we already enabled and causing issues. Running just `System.Data.SqlClient` or just `Microsoft.Data.SqlClient` test under net461 always worked ok.

I upgraded the MDS nuget to the latest version (2.1.0) and now with the support of the new EventSource all tests pass.

I did a CHANGELOG.md update and a few minor comment clarifications and fixes. Still need to add a few more new tests and update README.

mbakalov commented:
Highlights of changes:

* `SqlClientInstrumentationOptions` now has different public API on .NET Core and .NET Framework
* Made `SqlClientDiagnosticListener` .NET Core-only (it was referencing the options properties no longer available on .NET Framework)
  * For that had to refactor some common properties into a new `SqlActivitySourceHelper`
  * DiagnosticListener is no longer created/used under .NET Framework
* On .NET Framework the new `SetStatementText` is **true by default** and will capture both sp and sqltext if available. Not sure if good idea or not.
* SqlEventSourceListener is now listening to both event sources

cijothomas commented on src/OpenTelemetry.Instrumentation.SqlClient/SqlClientInstrumentationOptions.cs:
```diff
@@ -39,6 +39,17 @@ public class SqlClientInstrumentationOptions
+        /// add the text of the executed Sql commands as the <see cref="SemanticConventions.AttributeDbStatement"/> tag.
+        /// Default value: True.
+        /// </summary>
+        public bool SetStatementText { get; set; } = true;
```
Lets keep the default `false`. Its a bit risky to collect full sql statement by default.

cijothomas replied:
Also - we can make it more obvious here that the statement text could be just storedproc name or the actual sql query.

cijothomas commented:
@mbakalov Thanks a lot!
LGTM overall. except we need to keep default as false. Given this bug fix requires API change, its okay to be merged before 1.0.0. We need 2 approvals as we are close to 1.0.0. @CodeBlanch - please share your review.

ALso, open to any other suggestion for the names for these:
.NET Fx
SetStatementText 

.NET core
SetStoredProcedureCommandName 
SetTextCommandContent

mbakalov commented:
Thanks @cijothomas, I wasn't sure about the default, will update to false!

mbakalov commented:
As for the property names, maybe:

* netfx
  * `SetCommandText`
* netcore
  * `SetStoredProcedureCommandText`
  * `SetTextCommandText`

mbakalov commented:
Changed the default to `false`. Added more comments and warnings just in case. Let me know if need to update/reword anything there to make more clear.

cijothomas commented:
For namings, alternate proposal:

netfx
SetDbStatement 
netcore
SetDbStatementForStoredProcedure
SetDbStatementForText

"Sets DBStatement semantic convention" can be the general comment for all, as its coming directly from spec. Then specifically clarify `SetDbStatement` could be SPName or Text. `SetDbStatementForStoredProcedure` and `SetDbStatementForText` can be similarly clarified.

CodeBlanch commented on src/OpenTelemetry.Instrumentation.SqlClient/Implementation/SqlActivitySourceHelper.cs:
```diff
@@ -0,0 +1,37 @@
+        public const string ActivitySourceName = "OpenTelemetry.SqlClient";
+        public const string ActivityName = ActivitySourceName + ".Execute";
+
+        public const string MicrosoftSqlServerDatabaseSystemName = "mssql";
```
nit: Might be better to move this definition to semantic conventions, that's where it came from (https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/database.md#connection-level-attributes).

mbakalov replied:
Thanks, I'll update this! Other things ("db.statement") are being pulled from semantic conventions so better to be consistent.

CodeBlanch commented on test/OpenTelemetry.Instrumentation.SqlClient.Tests/OpenTelemetry.Instrumentation.SqlClient.Tests.csproj:
```diff
@@ -17,7 +17,7 @@
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.Data.SqlClient" Version="1.1.1" Condition="'$(TargetFramework)' != 'net452'" />
+    <PackageReference Include="Microsoft.Data.SqlClient" Version="2.1.0" Condition="'$(TargetFramework)' != 'net452'" />
```
Should we target 2.0.0? Don't want to force an unnecessary upgrade on anyone.

cijothomas replied:
This is test right?

mbakalov replied:
Yeah this is in the Tests csproj. The instrumentation itself doesn't reference it. For Tests I just selected to the latest stable on NuGet. Probably ok to keep?

cijothomas replied:
i mean - the actual project shouldn't depend on any package at all. For tests, we can test with latest, and it should be fine. Right?

CodeBlanch replied:
My bad! Fine to keep.

CodeBlanch reviewed (approved):
Left a couple of comments, but LGTM.

cijothomas commented:
Good to merge as has enough approvals, and no blocking comments. We just need to see if any better names are possible for the options.
